flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,270 @@
+                int index = getSlotRoundRobinIndex(entry.getValue().size(), ssg);
+
+                for (SchedulingExecutionVertex sev : entry.getValue()) {
+                    // For no CoLocations or with CoLocations with max parallelism of the SSG.
+                    if (!coLocationGroupMap.containsKey(entry.getKey())
+                            || ssgToExecutionSSGs.get(ssg).size() == entry.getValue().size()) {
+                        index = addVertexToExecutionSlotSharingGroup(executionSSGs, index, sev);
+                        continue;
+                    }
+
+                    // For CoLocations with exited execution slot sharing group.
+                    final CoLocationConstraint clc = getCoLocationConstraint(entry.getKey(), sev);
+                    if (clcToESsgMap.get(clc) != null) {
+                        clcToESsgMap.get(clc).addVertex(sev.getId());
+                        executionSlotSharingGroupMap.put(sev.getId(), clcToESsgMap.get(clc));
+                        continue;
+                    }
+                    // For CoLocations without existed execution slot sharing group.
+                    index = findAvailableESSGForCoLocation(sev, executionSSGs);
+                    clcToESsgMap.put(clc, executionSSGs.get(index));
+                    index = addVertexToExecutionSlotSharingGroup(executionSSGs, index, sev);
+                }
+                updateSlotRoundRobinIndexIfNeeded(entry.getValue().size(), ssg, index);
+            }
+            return executionSlotSharingGroupMap;
+        }
+
+        private void computeSsgToESsgMap(
+                LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> allVertices) {
+
+            allVertices.entrySet().stream()
+                    .map(
+                            entry ->
+                                    new AbstractMap.SimpleEntry<>(
+                                            entry.getKey(), entry.getValue().size()))
+                    .collect(
+                            Collectors.groupingBy(entry -> slotSharingGroupMap.get(entry.getKey())))
+                    .forEach(
+                            (ssg, entries) -> {
+                                Integer slotNumOfSsg =
+                                        entries.stream()
+                                                .map(Map.Entry::getValue)
+                                                .max(Comparator.comparingInt(o -> o))
+                                                .orElseThrow(
+                                                        () ->
+                                                                new FlinkRuntimeException(
+                                                                        String.format(
+                                                                                "Error state when computing the number of slots in %s",
+                                                                                ssg)));
```

Comment:
```suggestion
                                                                        Tuple2.of(
                                            slotSharingGroupMap.get(entry.getKey()),
                                            entry.getValue().size()))
                    .collect(
                            Collectors.groupingBy(
                                    tuple -> tuple.f0,
                                    Collectors.summarizingInt(tuple -> tuple.f1)))
                    .forEach(
                            (ssg, statistics) -> {
                                int slotNumOfSsg = statistics.getMax();
```


How about this update? The agg just needs `ssg` and `parallelism`, so:
- first step: generate ssg and parallelism.
- second step: compute max by `summarizingInt`
- get max directly. In theory, the option shouldn't be empty, so the `summarizingInt` is easy to use (It doesn't return Optional)